### PR TITLE
Remove openjdk8 and replace with openjdk11

### PR DESCRIPTION
**Pull request by anphel31**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
Remove toolchain code to build jdk8 from scratch. Modify SPECS to compile with a downloaded JDK11 RPM.

###### Change Log  

- Remove openjdk8 SPEC files
- Remove all toolchain sources and code required to compile openjdk8 from scratch
- Modify SPECs to compile with JDK11
- Download msopenjdk-11 RPM and include by default in the toolchain pkggen_core files.
- Remove ant-contrib which is no longer required and currently only builds with JDK8.
- Update validation scripts to allow the msopenjdk-11 RPM, which does not have a corresponding SPEC since it is built outside of CBL-Mariner.

###### Does this affect the toolchain?  

**YES**

###### Test Methodology

- Pipeline build id: xxxx

### CODE REVIEW

File: SPECS/R/R.spec [hunk 1]
```diff
@@ -120,6 +117,8 @@ TZ="Europe/Paris" make check -k -i
 %endif
 
 %changelog
+* Thu Dec 02 2021 Andrew Phelps <anphel@microsoft.com> - 4.1.0-2
```
Comment (anphel31) [inline SPECS/R/R.spec:120]: Note: the release was already set to 2 (mistakenly)

File: SPECS/ant/ant.spec [hunk 2]
```diff
@@ -14,32 +14,32 @@
 # published by the Open Source Initiative.
 #
 
-%global debug_package %{nil}
-%global ant_home %{_datadir}/ant
+%global debug_package   %{nil}
+%global ant_home        %{_datadir}/ant
+%global _mavenpomdir    /usr/share/maven-poms
```
Comment (thcrain-msft) [inline SPECS/ant/ant.spec:19]: Nit: `/usr/share` could be a macro

File: SPECS/ant/ant.spec [hunk 3]
```diff
@@ -14,32 +14,32 @@
...
+
 Summary:        Apache Ant
 Name:           ant
-Version:        1.10.9
-Release:        7%{?dist}
+Version:        1.10.11
```
Comment (thcrain-msft) [inline SPECS/ant/ant.spec:23]: Missing changelog entry for this file

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh [hunk 4]
```diff
@@ -326,26 +326,17 @@ chroot_and_install_rpms freetype
...
     ;;
     aarch64)
-        echo $($LFS/usr/lib/jvm/OpenJDK-1.8.0.181-bootstrap/bin/java -version)
-        build_rpm_in_chroot_no_install openjdk8_aarch64
+        wget -nv --timeout=30 https://packages.microsoft.com/cbl-mariner/2.0/preview/Microsoft/aarch64/msopenjdk-11-11.0.13%2B8-LTS-4.aarch64.rpm --directory-prefix=$CHROOT_RPMS_DIR_ARCH
     ;;
 esac
 
-# Install OpenJDK and OpenJRE
-chroot_and_install_rpms openjdk8
```
Comment (thcrain-msft) [inline toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:344]: Out of curiosity, do you know why we installed it before? Just want to make sure nothing was implicitly depending on this being installed.

File: toolkit/scripts/toolchain/create_toolchain_in_container.sh [hunk 5]
```diff
@@ -18,9 +18,6 @@ docker ps -a
 # docker rmi $(docker images -a -q)
 # docker rmi $(docker history marinertoolchain -q)
 
-# OpenJDK8 patch files
-cp -v $MARINER_SPECS_DIR/openjdk8/Awt_build_headless_only.patch ./container
```
Comment (thcrain-msft) [inline toolkit/scripts/toolchain/create_toolchain_in_container.sh:22]: I believe these files have corresponding entries in `toolkit/scripts/.gitignore`, can we remove those as well?

Comment (anphel31) [reply toolkit/scripts/toolchain/create_toolchain_in_container.sh:22]: Sure thing - will remove

File: SPECS/ant/ant.spec (hunk 3 above)
Comment (anphel31) [reply SPECS/ant/ant.spec:23]: Thanks, will add

File: SPECS/ant/ant.spec (hunk 2 above)
Comment (anphel31) [reply SPECS/ant/ant.spec]: fixed

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh (hunk 4 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:344]: I believe we initially built all of the jdk-dependent packages in the toolchain. I don't believe anything currently in the toolchain has any implicit dependencies though.

File: cgmanifest.json [hunk 6]
```diff
@@ -55,8 +55,8 @@
         "type": "other",
```
Comment (thcrain-msft) [inline cgmanifest.json:55]: Do we need a cgmanifest entry for the downloaded openjdk? Also, did we ever track the removed OpenJDK raw toolchain dependencies here?

Comment (anphel31) [reply cgmanifest.json:55]: Thanks for the reminder. For openjdk raw dependencies these were added in #1096 . Will remove these

Comment (anphel31) [reply cgmanifest.json:55]: And yes I think that the downloaded openjdk needs to be in cgmanifest as well. This has been added.

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh (hunk 4 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:344]: Confirmed by looking through our old repo. We initially built apache-maven in the toolchain which required the jdk8 install

**Review by thcrain-msft — APPROVED** [review]

File: toolkit/scripts/toolchain/container/toolchain-remote-wget-list [hunk 7]
```diff
@@ -64,71 +64,6 @@ https://www.openssl.org/source/openssl-1.1.1g.tar.gz
...
-https://www.x.org/pub/individual/lib/libXtst-1.2.3.tar.bz2
-https://www.x.org/pub/individual/lib/libXv-1.0.11.tar.bz2
-https://www.x.org/pub/individual/lib/libXvMC-1.0.10.tar.bz2
-https://www.x.org/pub/individual/lib/libXxf86dga-1.1.4.tar.bz2
-https://www.x.org/pub/individual/lib/libXxf86vm-1.1.4.tar.bz2
-https://www.x.org/pub/individual/lib/libdmx-1.1.3.tar.bz2
-https://www.x.org/pub/individual/lib/libpciaccess-0.13.5.tar.bz2
-https://www.x.org/pub/individual/lib/libxkbfile-1.0.9.tar.bz2
-https://www.x.org/pub/individual/lib/libxshmfence-1.2.tar.bz2
 https://download.savannah.gnu.org/releases/freetype/freetype-2.9.1.tar.gz
```
Comment (jslobodzian) [inline toolkit/scripts/toolchain/container/toolchain-remote-wget-list:67]: who needs fontconfig and freetype?  Can those be removed too?

**Review by jslobodzian — APPROVED** [review]

File: toolkit/scripts/toolchain/container/toolchain-remote-wget-list (hunk 7 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/container/toolchain-remote-wget-list:67]: No, fontconfig and freetype cannot be removed. There are a number of packages that still depend on them (cairo, harfbuzz, qt4-qtbase, pango, etc..)

Comment (anphel31) [reply toolkit/scripts/toolchain/container/toolchain-remote-wget-list]: Actually, you're right that these can be removed from the toolchain, since jdk8 was the only toolchain component that needed them. We just need to keep the specs for other package dependencies

File: cgmanifest.json [hunk 8]
```diff
@@ -4094,76 +3784,6 @@
...
-          "version": "3.5.13",
-          "downloadUrl": "https://www.x.org/pub/individual/lib/libXpm-3.5.13.tar.bz2"
-        }
-      }
-    },
-    {
-      "component": {
-        "type": "other",
-        "other": {
-          "name": "libXrandr",
```
Comment (PawelWMS) [inline cgmanifest.json:4121]: Why are we removing packages like these from the `cgmanifest.json`? I think we still have them. And is this related to the `openjdk` update?

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh [hunk 9]
```diff
@@ -316,36 +315,17 @@ build_rpm_in_chroot_no_install python3
 rm -vf $FINISHED_RPM_DIR/python3*debuginfo*.rpm
 chroot_and_install_rpms python3
 
-
-# openjdk needs fontconfig to build, which needs libxml
-build_rpm_in_chroot_no_install libxml2
```
Comment (PawelWMS) [inline toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:321]: Should we also remove `libxml2` packages from the toolchain manifest files?

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh [hunk 10]
```diff
@@ -316,36 +315,17 @@ build_rpm_in_chroot_no_install python3
...
-        echo $($LFS/usr/lib/jvm/OpenJDK-212-b04-bootstrap/bin/java -version)
-        build_rpm_in_chroot_no_install openjdk8
+        wget -nv --timeout=30 https://packages.microsoft.com/cbl-mariner/2.0/preview/Microsoft/x86_64/msopenjdk-11-11.0.13%2B8-LTS-4.x86_64.rpm --directory-prefix=$CHROOT_RPMS_DIR_ARCH
     ;;
     aarch64)
-        echo $($LFS/usr/lib/jvm/OpenJDK-1.8.0.181-bootstrap/bin/java -version)
-        build_rpm_in_chroot_no_install openjdk8_aarch64
+        wget -nv --timeout=30 https://packages.microsoft.com/cbl-mariner/2.0/preview/Microsoft/aarch64/msopenjdk-11-11.0.13%2B8-LTS-4.aarch64.rpm --directory-prefix=$CHROOT_RPMS_DIR_ARCH
     ;;
 esac
```
Comment (PawelWMS) [inline toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:327]: Suggestion: I think we can now drop the whole `case` statement and use this instead:

```bash
HOST_ARCH="$(uname -m)"
wget -nv --timeout=30 https://packages.microsoft.com/cbl-mariner/2.0/preview/Microsoft/"$HOST_ARCH"/msopenjdk-11-11.0.13%2B8-LTS-4."$HOST_ARCH".rpm --directory-prefix=$CHROOT_RPMS_DIR_ARCH
```

What do you think?

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh (hunk 9 above)
Comment (PawelWMS) [reply toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:321]: Nevermind, I see now it was simply moved a bit later in the script.

**Review by PawelWMS — COMMENTED** [review]
My main concern is the removal of some packages from `cgmanifest.json`. Everything else are just minor suggestions.

File: toolkit/tools/pkggen/worker/create_worker_chroot.sh [hunk 11]
```diff
@@ -32,7 +32,7 @@ install_one_toolchain_rpm () {
     fi
 
     echo "Found full path for package $1 in $rpm_path: ($full_rpm_path)" >> "$chroot_log"
-    rpm -i -v --nodeps --noorder --force --root "$chroot_builder_folder" --define '_dbpath /var/lib/rpm' "$full_rpm_path" &>> "$chroot_log"
+    rpm -i -v --nodeps --noorder --force --noscripts --root "$chroot_builder_folder" --define '_dbpath /var/lib/rpm' "$full_rpm_path" &>> "$chroot_log"
```
Comment (thcrain-msft) [inline toolkit/tools/pkggen/worker/create_worker_chroot.sh:35]: Can we special-case the `noscripts` flag to only apply to the msopenjdk RPM? I believe applying this to all RPMs will prevent things like our ldconfig scriptlets from running.

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh (hunk 9 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:321]: Yes, there are a number of packages that depend on libxml2. After seeing build errors I'm going to move it back to this spot.

File: cgmanifest.json (hunk 8 above)
Comment (anphel31) [reply cgmanifest.json:4121]: A number of the x.org dependencies were added to the toolchain in order to compile jdk8. Since that's being removed, so are the dependencies. Although it looks like some of these have been added as SPECS from the coreui repo. I'll double check and restore those case-by-case.

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh (hunk 10 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:327]: I have a slight preference for the case statement, since the URLs can be copied and used as-is without variable substitution.

File: toolkit/tools/pkggen/worker/create_worker_chroot.sh (hunk 11 above)
Comment (anphel31) [reply toolkit/tools/pkggen/worker/create_worker_chroot.sh]: good idea, I've updated the script to special-case this package

**Review by PawelWMS — APPROVED** [review]

File: toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go [hunk 12]
```diff
@@ -270,10 +270,15 @@ func (r *RpmRepoCloner) Clone(cloneDeps bool, packagesToClone ...*pkgjson.Packag
 			pkgName,
 		}
 
-		if cloneDeps {
-			args = append([]string{"download", "--alldeps"}, args...)
-		} else {
+		// special case to skip dependencies for msopenjdk-11
+		if (pkgName == "msopenjdk-11-11.0.13+8_LTS-4.x86_64") || (pkgName == "msopenjdk-11-11.0.13+8-LTS-4.aarch64") {
```
Comment (PawelWMS) [inline toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go:274]: With the current solution any change to the version name will cause us to miss this check. How about:

```suggestion
		if strings.HasPrefix(pkgName, "msopenjdk-") {
```

File: toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go [hunk 13]
```diff
@@ -270,10 +270,15 @@ func (r *RpmRepoCloner) Clone(cloneDeps bool, packagesToClone ...*pkgjson.Packag
...
 			args = append([]string{"download-nodeps"}, args...)
+		} else {
+			if cloneDeps {
+				args = append([]string{"download", "--alldeps"}, args...)
+			} else {
+				args = append([]string{"download-nodeps"}, args...)
+			}
```
Comment (PawelWMS) [inline toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go:281]: Suggestion to decrease the nesting of `if` statements:

```suggestion
		} else if cloneDeps {
			args = append([]string{"download", "--alldeps"}, args...)
		} else {
			args = append([]string{"download-nodeps"}, args...)
		}
```

File: toolkit/scripts/toolchain/check_manifests.sh [hunk 14]
```diff
@@ -36,8 +36,16 @@ write_rpms_from_toolchain () {
     specs=$(cat $TOOLCHAIN_SPEC_LIST_FILE)
     for specName in $specs
     do
-        specFile=$(find $SPECS_DIR/**/$specName.spec)
-        write_rpms_from_spec $specFile $1
+        if [[ $specName == *"msopenjdk"* ]]; then
```
Comment (MateuszMalisz) [inline toolkit/scripts/toolchain/check_manifests.sh:39]: nit:quotes
I would also suggest using `[[ "$specName" =~ ".*msopenjdk.*" ]]`. Do these asterisks expand globs there?

File: toolkit/scripts/toolchain/check_manifests.sh [hunk 15]
```diff
@@ -36,8 +36,16 @@ write_rpms_from_toolchain () {
...
+            # special case to add msopenjdk-11 which is downloaded and does not have a SPEC
+            jdkfilename=$(cat "$MANIFESTS_DIR/toolchain_$ARCH.txt" | grep $specName)
```
Comment (MateuszMalisz) [inline toolkit/scripts/toolchain/check_manifests.sh:41]: nit: no need to cat, grep can operate on the file.

File: toolkit/scripts/toolchain/check_manifests.sh [hunk 16]
```diff
@@ -36,8 +36,16 @@ write_rpms_from_toolchain () {
...
+            echo "adding special case for: $jdkfilename"
+            echo "$jdkfilename" >> $1
+        else
+            # normal case - search for the RPM filename using the SPEC name
+            specFile=$(find $SPECS_DIR/**/$specName.spec)
```
Comment (MateuszMalisz) [inline toolkit/scripts/toolchain/check_manifests.sh:46]: nit: isn't it always `$SPECS_DIR/$specName/$specName.spec`?

File: toolkit/tools/pkggen/worker/create_worker_chroot.sh [hunk 17]
```diff
@@ -31,8 +31,15 @@ install_one_toolchain_rpm () {
         exit 1
     fi
 
+    if [[ $1 == *"msopenjdk"* ]]; then
```
Comment (MateuszMalisz) [inline toolkit/tools/pkggen/worker/create_worker_chroot.sh:34]: nit: quotes
Comment about =~above applies as well

File: toolkit/scripts/toolchain/check_manifests.sh (hunk 16 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/check_manifests.sh:46]: not always. for example we used to have openjdk8/openjdk8.spec and openjdk8/openjdk8_aarch64.spec

File: toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go (hunk 12 above)
Comment (anphel31) [reply toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go:274]: thanks

File: toolkit/scripts/toolchain/check_manifests.sh (hunk 15 above)
Comment (anphel31) [reply toolkit/scripts/toolchain/check_manifests.sh]: fixed

File: toolkit/tools/pkggen/worker/create_worker_chroot.sh (hunk 17 above)
Comment (anphel31) [reply toolkit/tools/pkggen/worker/create_worker_chroot.sh]: sure; updated

File: toolkit/tools/srpmpacker/srpmpacker.go [hunk 18]
```diff
@@ -305,8 +305,13 @@ func findSPECFiles(specsDir string, packList []string) (specFiles []string, err
 				return
 			}
 			if len(specFile) != 1 {
-				err = fmt.Errorf("unexpected number of matches (%d) for spec file (%s)", len(specFile), specName)
-				return
+				if( specName == "msopenjdk-11" ) {
+					logger.Log.Infof("Ignoring missing match for (%s)", specName)
+					continue
+				} else {
```
Comment (PawelWMS) [inline toolkit/tools/srpmpacker/srpmpacker.go:311]: Here I would also try to be more flexible with the version and logging, since it might be confusing for the user:

```suggestion
				if (strings.HasPrefix(specName, "msopenjdk-11" ) {
					logger.Log.Debug("Ignoring missing match for the 'msopenjdk' package, which is externally-provided and thus doesn't have a local spec.")
					continue
				} else {
```

Comment (anphel31) [reply toolkit/tools/srpmpacker/srpmpacker.go]: sounds good, will update
